services/audit-service/src/main/java/datawave/microservice/audit/auditors/file/FileAuditor.java

```diff
@@ -0,0 +1,212 @@
+import java.io.OutputStream;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLEncoder;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * An implementation for {@link Auditor}, which writes URL encoded, JSON formatted audit messages to a file.
+ */
+public class FileAuditor implements Auditor {
+    
+    private final Logger log = LoggerFactory.getLogger(this.getClass());
+    
+    protected static final ObjectMapper mapper = new ObjectMapper();
+    
+    protected final SimpleDateFormat sdf;
+    protected final ReentrantLock writeLock = new ReentrantLock(true);
+    
+    protected long maxFileLenBytes;
+    protected long maxFileAgeMillis;
+    
+    protected FileSystem fileSystem;
+    protected Path path;
+    
+    protected Path currentFile = null;
+    protected Date creationDate = null;
+    
+    protected FileAuditor(Builder<?> builder) throws URISyntaxException, IOException {
+        this.maxFileLenBytes = builder.maxFileLenBytes;
+        this.maxFileAgeMillis = builder.maxFileAgeMillis;
+        
+        Configuration config = new Configuration();
+        
+        if (builder.configResources != null) {
+            for (String resource : builder.configResources)
+                config.addResource(new Path(resource));
+        }
+        
+        path = new Path(builder.path);
+        
+        if (builder.subpath != null)
+            path = new Path(path, builder.subpath);
+        
+        if (builder.prefix != null)
+            path = new Path(path, builder.prefix);
+        
+        fileSystem = FileSystem.get(path.toUri(), config);
+        
+        if (!fileSystem.exists(path))
+            fileSystem.mkdirs(path);
+        
+        this.sdf = new SimpleDateFormat("yyyyMMdd_HHmmss.SSS'.json'");
+    }
+    
+    @Override
+    public void audit(AuditParameters auditParameters) throws Exception {
+        
+        // convert the messages to URL-encoded JSON
+        Map<String,String> auditParamsMap = auditParameters.toMap();
+        auditParamsMap.forEach((key, value) -> auditParamsMap.put(key, urlEncodeString(value)));
+        String jsonAuditParams = mapper.writeValueAsString(auditParamsMap) + "\n";
+        
+        writeLock.lock();
+        try {
+            // if the file/stream is null, doesn't exist, or the file is too old/big, create a new file & output stream
+            if (currentFile == null || !fileSystem.exists(currentFile) || isFileTooOld() || isFileTooBig())
+                createNewFile();
+            
+            writeAudit(jsonAuditParams);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+    
+    protected String urlEncodeString(String value) {
```

Comment:
Why the need for this? If the ObjectMapper won't do the URL encoding for you, then what's the purpose of catching the exception and returning the original value? If that case happens, then the object mapper will just fail on it. If the object mapper does take care of the encoding, then there's no need for this method.

Reply:
I'm just thinking ahead to the moment when I will be reading these json-formatted files in spark.  Spark natively supports reading json-formatted data, where each json object is written on a single line.

Multi-line json objects are also supported, but you have to write all of the json objects as an array, which isn't as convenient when writing in append mode (basically, it would be nice to avoid having to write beginning & closing square brackets).  

To get around this, I was planning on just URL-encoding the parameter values to ensure that no newlines would be present in a given json object.  This is easy to undo in spark when it comes time to process the data.  

To my knowledge, you don't just get URL encoding for free by using an object mapper, so that's why I have the extra step.  Also, while I do have logic for catching an unsupported encoding exception, you can see that it's hard coded to UTF8 (which is a supported character encoding), so I don't expect that exception to ever be thrown.  

Alternatively, I suppose I could write Base64 encoded bytes instead and URL encoding altogether, but i thought this would be good enough.  

What are your thoughts?

Reply:
The URL encoding sounds a like hoky to me.  In fact any nested encoding makes the file difficult to parse in other ways.  I would rather create an appropriate json output that can be readable by more than the spark processing.

Reply:
Sounds like some sort of escaping is required in any case, but using a standard JSON writer would avoid having to do anything like this.  I understand your Spark concerns, but I would rather have standardized json output that Spark can handle instead of having to do a urlDecode wherever we have to read these files.

Reply:
I had assumed that I would need to URL encode to ensure that the json stays as a single line (to play nicely with Spark), but after testing out the object mapper without doing any URL encoding, I think that will work just fine.  I updated the pr.